Update Durable Nonces Course

darkvallen opened the pull request:
### Problem



### Summary of Changes
- Updated code snippets
- Use Partial signing for transactions with multiple signers.
- Use `getExplorerLink` from solana-helpers for easier link generation.
- Added the calculation for the exact rent-exempt balance using `connection.getMinimumBalanceForRentExemption`.
- Added screenshots of successful test runs are included in the documentation.
- Fixed grammar and styling
- Fixed as per guidelines


Fixes #
Unboxed PRs
https://github.com/Unboxed-Software/solana-lab-durable-nonces/pull/1
<!-- 
Note: The maintainers of this repo may make editorial changes as needed without creating comments.
Please ensure you have "Allow edits by maintainers" setting on this PR enabled to help speed up the review process. Thanks :)
-->

mikemaccana commented on content/courses/offline-transactions/durable-nonces.md:
```diff
@@ -14,39 +14,39 @@ description: "Use durable nonces to sign transactions ahead of time."
-   the required singers/authorities. For instance, if you are transferring SOL
-   from your wallet to another, you'll need to sign the transaction so the
-   network can verify that the transaction is valid.
+2. **Signatures**: Signatures are proof that the transaction was signed by the
+   required signers/authorities. For instance, if you are transferring SOL from
+   your wallet to another, you'll need to sign the transaction so the network
+   can verify that the transaction is valid.
```
This is a criticism of the old writing, but the logic here is circular and bad. Signatures prove I sign it, which means I have a signature, which proves I signed it. 

Try:

> **Signatures**: Signatures are proof that the transaction was made by the required signers (assuming they have kept their secret keys secure).

mikemaccana commented on content/courses/offline-transactions/durable-nonces.md:
```diff
@@ -14,39 +14,39 @@ description: "Use durable nonces to sign transactions ahead of time."
 Durable Nonces are a way to bypass the expiration date of regular transactions.
-To understand that better, we'll start by looking at the concepts behind regular
-transactions.
+To understand this more clearly, we'll start by looking at the concepts behind
+regular transactions.
```
```suggestion
To understand this, we'll start by looking at the concepts behind
regular transactions.
```

mikemaccana replied:
Don't be afraid to clean up unnecessarily wordy language.

mikemaccana commented on content/courses/offline-transactions/durable-nonces.md:
```diff
@@ -14,39 +14,39 @@ description: "Use durable nonces to sign transactions ahead of time."
+regular transactions.
 
-In Solana, transactions are made of three main parts:
+In Solana, transactions consist of three main parts:
```
```suggestion
In Solana, transactions have three main parts:
```

Plain language.

mikemaccana commented on content/courses/offline-transactions/durable-nonces.md:
```diff
@@ -115,30 +115,30 @@ From the
-This is not meant to provoke hysteria, just as a PSA of what's possible. This is
-why you should only put into hot wallets what you're willing to lose and don't
-sign with your cold wallet.
+This is not meant to provoke hysteria, but serves as a PSA of what's possible.
```
```suggestion
This is not meant to provoke hysteria, but serves to show what's possible.
```

PSAs are US-specific.

mikemaccana replied:
This lesson goes from 'you are the developer' to 'you are the user'. We should fix that.

Maybe as developers, we tell people their durable nonce trandsactions may be flagged by wallets, and then explain that durable nonces are often used for malicious transactions?

darkvallen replied:
I have revised the section to maintain a consistent perspective, addressing both developers and users. The updated text now informs developers to educate users about the potential risks and how to mitigate them.

Please let me know if there are any further adjustments needed.

mikemaccana commented on content/courses/offline-transactions/durable-nonces.md:
```diff
@@ -115,30 +115,30 @@ From the
-Durable nonces are a way to sign transactions offchain and keep them in storage
-until they are ready to be submitted to the network. And this allows us to
-create durable transactions.
+Durable nonces are a way to sign transactions off-chain and keep them in storage
```
offchain is one word, please fix everywhere.

mikemaccana commented on content/courses/offline-transactions/durable-nonces.md:
```diff
@@ -150,72 +150,75 @@ It is important to note that durable nonces require
 Durable transactions differ from regular transactions in the following ways:
 
 1. Durable Nonces replace the recent blockhash with a nonce. This nonce is
    stored in a `nonce account` and will be used only once in one transaction.
```
Not sure `nonce account` is code or a file name. Maybe just 'nonce account' ?

mikemaccana reviewed (changes requested):
Great work, a few minor bits but it's very clear you've read the contributing guidelines. Much appreciated, let's get this merged!

darkvallen commented:
> Great work, a few minor bits but it's very clear you've read the contributing guidelines. Much appreciated, let's get this merged!

Updated, Could you please review the updated "Considerations" section and let me know if it addresses your concerns?

mikemaccana commented:
Marked as winner, congrats!

mikemaccana commented:
Also @darkvallen: for https://github.com/Unboxed-Software/solana-lab-durable-nonces can you please make your changes to the `solution` branch, not `new-solution`? Also please update `starter` branch if necessary.

darkvallen commented:
> Also @darkvallen: for https://github.com/Unboxed-Software/solana-lab-durable-nonces can you please make your changes to the `solution` branch, not `new-solution`? Also please update `starter` branch if necessary.
> 
> Reply here once done.

Thanks @mikemaccana! A new learning experience for me.

Done https://github.com/Unboxed-Software/solana-lab-durable-nonces/pull/4 here's the `solution` branch PR.
